### PR TITLE
bpo-37999: No longer use __int__ in implicit integer conversions.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Only `__index__` should be used to make integer conversions lossless.

https://bugs.python.org/issue37999

### CODE REVIEW

File: Doc/whatsnew/3.9.rst [hunk 1]
```diff
@@ -99,6 +99,13 @@ Other Language Changes
   ignored for empty strings.
   (Contributed by Victor Stinner in :issue:`37388`.)
 
+* Builtin and extension functions that take integer arguments no longer accept
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.9.rst]: TODO: Move it to 3.10.

File: Lib/datetime.py [hunk 2]
```diff
@@ -2539,7 +2508,7 @@ def _name_from_offset(delta):
     # Clean up unused names
     del (_DAYNAMES, _DAYS_BEFORE_MONTH, _DAYS_IN_MONTH, _DI100Y, _DI400Y,
          _DI4Y, _EPOCH, _MAXORDINAL, _MONTHNAMES, _build_struct_time,
-         _check_date_fields, _check_int_field, _check_time_fields,
+         _check_date_fields, _index, _check_time_fields,
```
Comment (mdickinson) [inline Lib/datetime.py]: Looks like there's an effort to keep things in alphabetical order here; maybe move `_index` to the right place with respect to that order?

File: Lib/test/test_math.py [hunk 3]
```diff
@@ -499,17 +499,13 @@ def testFactorial(self):
             self.assertEqual(math.factorial(i), total)
             self.assertEqual(math.factorial(i), py_factorial(i))
         self.assertRaises(ValueError, math.factorial, -1)
-        self.assertRaises(ValueError, math.factorial, -10**100)
+        self.assertRaises((ValueError, OverflowError), math.factorial, -10**100)
```
Comment (mdickinson) [inline Lib/test/test_math.py]: Hmm. Getting an `OverflowError` here seems wrong to me, but perhaps that's something that could be fixed independently of this PR.

**Review by mdickinson — APPROVED** [review]
LGTM; I'm happy to see this finally reach completion.

~Not sure I like the `OverflowError` for large negative values for `math.factorial`: that seems like a case of the implementation details leaking into the API.~  EDIT: This has been changed

Just to be clear, this PR doesn't change the fact that `PyNumber_Index` can still return something that's an instance of a strict subclass of `int`. Is that correct?

**Comment by serhiy-storchaka** [comment]
> Just to be clear, this PR doesn't change the fact that `PyNumber_Index` can still return something that's an instance of a strict subclass of `int`. Is that correct?

Yes, it is correct.

File: Lib/test/test_math.py (hunk 3 above)
Comment (mdickinson) [reply Lib/test/test_math.py]: Resolved. Thank you!

**Pull request merged**